synapse/util/caches/descriptors.py

```diff
@@ -53,9 +53,9 @@
 
 
 class CachedFunction(Generic[F]):
-    invalidate: Any = None
-    invalidate_all: Any = None
-    prefill: Any = None
+    invalidate: Callable[[Tuple[Any, ...]], None]
```

Comment:
Thanks for fixing this!

For my own reference, related to https://github.com/matrix-org/synapse/pull/13863 where we weren't invalidating the `have_seen_event` cache as expected but that was because we weren't putting keys in the cache the right way.

The first pass there adjusted the `invalidate` usage but that was just masking the underlying problem and this typing would have caught that.

And just generally helps confidence that we're calling this correctly all over the place!!